common/cache.go

```diff
@@ -0,0 +1,112 @@
+package common
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/zcash-hackworks/lightwalletd/walletrpc"
+)
+
+type BlockCacheEntry struct {
+	data []byte
+	hash []byte
+}
+
+type BlockCache struct {
+	MaxEntries int
+
+	FirstBlock int
+	LastBlock  int
+
+	m map[int]*BlockCacheEntry
+
+	mutex sync.RWMutex
+}
+
+func NewBlockCache(maxEntries int) *BlockCache {
+	return &BlockCache{
+		MaxEntries: maxEntries,
+		FirstBlock: -1,
+		LastBlock:  -1,
+		m:          make(map[int]*BlockCacheEntry),
+	}
+}
+
+func (c *BlockCache) Add(height int, block *walletrpc.CompactBlock) (error, bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.FirstBlock == -1 && c.LastBlock == -1 {
+		// If this is the first block, prep the data structure
+		c.FirstBlock = height
+		c.LastBlock = height - 1
+	}
+
+	// If we're adding a block in the middle of the cache, remove all
+	// blocks after it, since this might be a reorg, and we don't want
+	// Any outdated blocks returned
+	if height >= c.FirstBlock && height <= c.LastBlock {
+		for i := height; i <= c.LastBlock; i++ {
+			delete(c.m, i)
+		}
+		c.LastBlock = height - 1
+	}
```

Comment:
This function should fail with an error unless either `height >= c.FirstBlock && height <= c.LastBlock` or `height = c.LastBlock + 1`. This will make sure the cache doesn't reach any of several weird states:

- Some `nil` blocks in the cache (currently could be reached by adding a block at height > c.LastBlock + 1).
- Blocks from a mixture of reorg forks (currently could be reached by adding a block at height <  c.FirstBlock; the higher-height blocks won't be deleted)
- Blocks where the PrevHash doesn't match the previous block's hash.

AFAICT enforcing this constraint won't break the ingestor code (except for the edge case about `height < startHeight` I left a comment about there).